overcast/runner/__init__.py

```diff
@@ -413,16 +474,36 @@ def provision_step(self, details):
         for base_node_name, node_info in stack['nodes'].items():
             def _create_node(base_name):
                 node_name = self.add_prefix(base_name)
-                self.nodes[base_name] = self.create_node(node_name, node_info,
-                                                         keypair=keypair_name,
-                                                         userdata=userdata)
+                self.nodes[base_name] = Node(node_name, node_info,
+                                                        keypair=keypair_name,
+                                                        userdata=userdata,
+                                                        runner=self)
+                self.nodes[base_name].build()
 
             if 'number' in node_info:
                 count = node_info.pop('number')
                 for idx in range(1, count+1):
-                    _create_node('%s%d' % (base_node_name, idx))
+                    node_name = '%s%d' % (base_node_name, idx)
+                    _create_node(node_name)
+                    pending_nodes.add(node_name)
             else:
                 _create_node(base_node_name)
+                pending_nodes.add(base_node_name)
+
+        while pending_nodes:
+            time.sleep(5)
+            for name in pending_nodes:
+                if self.nodes[name].poll() == 'ACTIVE':
+                    done.add(name)
+                if self.nodes[name].poll() == 'ERROR':
+                    if details.get('retry-if-fails', False):
+                        wait()
+                        if deadline and time.time() > deadline:
+                            raise exceptions.ProvisionTimedOutException(stdin)
+                        self.nodes[name].clean()
+                        self.nodes[name].build()
+                    raise exceptions.ProvisionFailedException(stdin)
+            pending_nodes = pending_nodes.difference(done)
```

Comment:
I think I'd change the loop logic a little bit here. Reason: I don't want to start by waiting 5 seconds.... but I also don't want to wait 5 seconds after everything is done, so just moving the sleep to the end of the loop won't be any better.

So, at the bottom of the loop, I'd break out of it when there are no more pending nodes, and otherwise sleep for N seconds:

    if not pending_nodes:
        break
    time.sleep(5)

...and since this is now where I decide to break the loop, I'd change the loop conditional to `while True`.